### ISSUE TITLE
`nix profile` has bad UX

### ISSUE THREAD

**Issue by bobvanderlinden** [NixOS/nix#7966]
There are quite a few UX problems with `nix profile`. I'd like to document them to give some context:

- [x] https://github.com/NixOS/nix/issues/7530
- [ ] https://github.com/NixOS/nix/issues/7960
- [ ] https://github.com/NixOS/nix/issues/7961
- [ ] https://github.com/NixOS/nix/issues/7962
- [ ] https://github.com/NixOS/nix/issues/7963
- [ ] https://github.com/NixOS/nix/issues/7964
- [ ] https://github.com/NixOS/nix/issues/7965
- [ ] https://github.com/NixOS/nix/pull/6481

I thought I'd gather them here in a single overarching issue, instead of listing the problems all here.

**Comment by bobvanderlinden** [comment]
3 of these issues can be resolved if we could refer to installed packages by name instead of by index:

- [ ] https://github.com/NixOS/nix/issues/7967

**Comment by thufschmitt** [comment]
Thanks a lot for listing all these. I think that will be incredibly useful for #7701

**Comment by alper** [comment]
I just got here and with all the documentation around it I still haven't been able to figure out what `nix profile` does and why I need it or how to use it.

Given how confusing a lot of the documentation is… maybe a full revamp is in order.

**Comment by iFreilicht** [comment]
#6481 was abandoned in favor of #8631; you may want to update the todo list :)

**Comment by viperML** [comment]
Thanks for compiling the issues. I wanted to write a home-manager alternative ( https://github.com/viperML/activation-manager ), that relies on being installed as part of a `buildEnv` + `nix profile install /flake#package-output`. But I just faced all the bad UX of nix profile.

**Comment by iFreilicht** [comment]
@bobvanderlinden now that #8678 has been merged, could you re-evaluate whether which of these issues are resolved?

**Comment by iFreilicht** [comment]
Two more issues that might be worth adding here: https://github.com/NixOS/nix/issues/10463, https://github.com/NixOS/nix/issues/10419

**Comment by bobvanderlinden** [comment]
@iFreilicht I think they can both be considered duplicates. I've commented in the issues.

**Comment by trim21** [comment]
`nix profile remove --all` will also remove `nix` itself, is this expected?

**Comment by iFreilicht** [comment]
@trim21 Yes. `nix profile` expects a system-wide nix installation to be present. If you end up in this situation, you can still search for nix by running a glob and picking one of the outputs. For example:

```console
# echo /nix/store/*-nix-2*/bin
/nix/store/009hyb1ryy8jrypqikcaz4nfh3xkbaaz-nix-2.18.2/bin/ /nix/store/h16lidvhs90r2pvqns0kkv9ffha0h25p-nix-2.19.4/bin/ /nix/store/hghqbpw1d5a2kqzjjn68wm9klqbmjvk0-nix-2.18.2/bin/ /nix/store/l876hksjiz1240d5in5ksdzskag0w15k-nix-2.18.7/bin/ /nix/store/ma0p24nzjcylflyn4bz4wj8kf28a7ida-nix-2.18.5/bin/ /nix/store/nxk53ss4jcbf6i3pf1r1gz7di5jhaajm-nix-2.18.2/bin/ /nix/store/nzf1wy0z8j8dpsv49x8g5i93c7i6bcph-nix-2.18.5/bin/ /nix/store/xcdazpgbpc3inr2hqsxahb0wmbndp04a-nix-2.19.6/bin/ /nix/store/ycn722fb286lpg0h32lg935l6nxjhkq5-nix-2.19.6/bin/
# /nix/store/009hyb1ryy8jrypqikcaz4nfh3xkbaaz-nix-2.18.2/bin/nix profile install nixpkgs#nix
# nix --version
nix (Nix) 2.18.7
```

If you think this is a bad idea and want to it to be changed, please open a separate issue so we can discuss it there :)

**Comment by elikoga** [comment]
https://github.com/NixOS/nix/issues/12380 should also be relevant
